### PR TITLE
Add prettyPrint option to logger type definitions

### PR DESCRIPTION
**Pull request by agramian**
The `FastifyLoggerOptions` interface is missing a type definition for the `prettyPrint` option which `pino` supports when `pino-pretty` is installed.  The latest general [Logging docs](https://www.fastify.io/docs/latest/Logging/) already reference the `prettyPrint` option however without the type definition an error is shown when using TypeScript and passing the option to the logger when creating the Fastify instance. 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: types/logger.d.ts
Comment (zekth) [inline types/logger.d.ts:64]: This is incomplete as the definition can either be a `boolean` or a `pino-pretty` option object. Please see: https://github.com/pinojs/pino-pretty#options

**Review by zekth — CHANGES_REQUESTED** [review]

**Comment by kibertoad** [comment]
Could you also add type test for this? See https://github.com/fastify/fastify/tree/master/test/types

File: types/logger.d.ts
Comment (agramian) [reply types/logger.d.ts:64]: Good point.  I will update.

**Comment by agramian** [comment]
> Could you also add type test for this? See https://github.com/fastify/fastify/tree/master/test/types

Yes.  Will do.

**Review by zekth — APPROVED** [review]
LGTM

File: types/logger.d.ts
Comment (L2jLiga) [inline types/logger.d.ts:4]: I'm a bit worried about TypeScript users. Will it work when user does not have `@types/pino` package installed?

Comment (zekth) [reply types/logger.d.ts:4]: Indeed!

Comment (L2jLiga) [reply types/logger.d.ts:4]: `pino` does not provides own types and they shipped via `@types/pino` package hence If I won't install `@types/pino` TypeScript will consider typings as invalid

Comment (agramian) [reply types/logger.d.ts:4]: So would you prefer to just copy the type definition over?  That way would mean it would have to be updated here on changes from `pino` / `@types/pino`.

Alternatively a note could be added to the docs.  In order to use `pino-pretty` with `fastify` you need to install the `pino-pretty` dependency regardless so in the docs a note can be added to also install `@types/pino` if you are using TypeScript.

Comment (L2jLiga) [reply types/logger.d.ts:4]: I would include typings in `pino` itself.
BTW moving `@types/pino` to `dependencies` or adding notes to README also acceptable, but as a WebStorm IDE user I would prefer first (moving `@types/pino`) for better smart completion.

Comment (kibertoad) [reply types/logger.d.ts:4]: If team agrees, I can submit PR for type inclusion in pino.

Comment (L2jLiga) [reply types/logger.d.ts:4]: I'm +1

Comment (fox1t) [reply types/logger.d.ts:4]: I don't like including types that are not under our control (from `@types/pino`) to the fastify dependency tree.
@mcollina what do you think? What about a PR to pino itself to add typings? @jsumners any thoughts?

Comment (Eomm) [reply types/logger.d.ts:4]: I think this has been already discussed in past and @fox1t get the point

Here one quick answer:
https://github.com/fastify/fastify/pull/1532#discussion_r265271715

Comment (Ethan-Arrowood) [reply types/logger.d.ts:4]: We have decided in the past not to ship the pino types with our module. We'd like it if there were instructions to install pino types if the user plans on using the pino logger. Otherwise, the logger types we provide should be the minimum types necessary for using the logger as is.

Comment (fox1t) [reply types/logger.d.ts:4]: Ok. So now we have one remaining option: since pino is already a dependency of fastify, if pino had integrated types we could have them out of the box. Is any particular reason why pino relies on `@types/`.

Comment (mcollina) [reply types/logger.d.ts:4]: Pino has no official types. No one of the maintainers of pino has enough typescript skills to have types. The types on `@types` are unofficial and should not be used.

**Review by mcollina — CHANGES_REQUESTED** [review]
It should not depend on pino types, as stated multiple times in a lot of previous PRS on this topic. Maybe we can add a comment in the file?

File: types/logger.d.ts
Comment (fox1t) [reply types/logger.d.ts:4]: @mcollina I agree with you. Do you want me to contribute to pino types and maintain them in the repo itself? I think that there are other folks here that can help to maintain them since pino is THE logger of Fastify.

Comment (kibertoad) [reply types/logger.d.ts:4]: @fox1t Would be happy to volunteer for helping with official pino types.

Comment (zekth) [reply types/logger.d.ts:4]: @fox1t i can help too

Comment (fox1t) [reply types/logger.d.ts:4]: I don't like including types that are not under our control (from `@types/pino`) to the fastify dependency tree.
@mcollina what do you think? What about a PR to pino itself to add typings? @jsumners any thoughts?

Comment (fox1t) [reply types/logger.d.ts:4]: Any thoughts?

Comment (mcollina) [reply types/logger.d.ts:4]: I would open an issue on pino, offering to write and maintain them in the long run and see it from there.

Comment (fox1t) [reply types/logger.d.ts:4]: Done: https://github.com/pinojs/pino/issues/910

**Comment by mcollina** [comment]
This is easy to fix by lifting whatever parts of the pino typedefs is needed in fastify itself. it's easy and swift.

**Comment by Ethan-Arrowood** [comment]
I'm not interested in maintaining more types than Fastify's so I'd rather we keep them here. Copy-pasting pino types and shipping them ourselves from within Fastify is okay as the pino api is not too complex. 

What I've tried to articulate in the past is that there exists a minimal logger api that most fastify users use. When a dev needs more than that minimal interface; we should recommend them to install the pino types and use them in their fastify project themselves. I can't remember if they need to use declaration merging or use a generic to achieve that. This is the best path that doesn't add any additional maintenance responsibility to the projects in question

**Comment by fox1t** [comment]
> This is easy to fix by lifting whatever parts of the pino typedefs is needed in fastify itself. it's easy and swift.

If we are fine in "duplicating" this over here, we should maintain Pino's typings in Fastify's codebase.

**Comment by fox1t** [comment]
> I'm not interested in maintaining more types than Fastify's so I'd rather we keep them here. Copy-pasting pino types and shipping them ourselves from within Fastify is okay as the pino api is not too complex.
> 
> What I've tried to articulate in the past is that there exists a minimal logger api that most fastify users use. When a dev needs more than that minimal interface; we should recommend them to install the pino types and use them in their fastify project themselves. I can't remember if they need to use declaration merging or use a generic to achieve that. This is the best path that doesn't add any additional maintenance responsibility to the projects in question

We the types we write diverge in any non-mergiable way from `@types/pino`, basically we creating conflicts to our users that might be difficult to spot.

If we are all fine with this, then I vote too to add them to Fastify. :)

**Comment by Ethan-Arrowood** [comment]
> We the types we write diverge in any non-mergiable way from @types/pino 

this is why I think we should only ship a minimal viable type api for the logger. if the user wants full pino types they should be instructed to `npm i -D @types/pino` and then pass the `Pino` type to their `FastifyInstance`

**Comment by fox1t** [comment]
> > We the types we write diverge in any non-mergiable way from @types/pino
> 
> this is why I think we should only ship a minimal viable type api for the logger. if the user wants full pino types they should be instructed to `npm i -D @types/pino` and then pass the `Pino` type to their `FastifyInstance`

Are you suggesting adding a new generics to Fastify typings? 🤔

If so, we might have conflicts anyway if both types are decalred using the same `namespace` - `interface name`.

**Comment by Ethan-Arrowood** [comment]
https://github.com/fastify/fastify/blob/master/types/instance.d.ts#L20 Generic already exists for users to pass their own logger instance (i.e. Pino). There is even a test case: https://github.com/fastify/fastify/blob/master/test/types/logger.test-d.ts#L46

One aspect I would like to see changed in v4 is that we move the logger generic to the front of the list so that users don't have to specify the server, request, and reply ones as well. (We could try to use a generic object like the requestgenericinterface but iirc that breaks some of the magical inference so it might not be an actual solution here).

**Comment by fox1t** [comment]
Now I am even more confused than before. I know you can already pass a generic to "inject" pino's typings, but this PR is trying to fix a different issue. Am I wrong?

**Comment by Ethan-Arrowood** [comment]
The inject method for pino typings is the solution to this issue. By recommending the user to inject pino typings they should get access to the prettyPrint property

**Comment by fox1t** [comment]
OK let me recap this for clarity, just to see if I understood what you are saying. 
You are suggesting is to NOT add a `prettyPrint` as a Fastify's factory option in the typings and to let the options being injected by Fastify user? Are we going to remove other typings from here https://github.com/fastify/fastify/blob/master/types/logger.d.ts#L43

If not, why some options are "integrated" and others must be provided by passing explicitly Pino typings?

Sorry for asking this again, but I have really hard time getting this. :)

**Comment by Ethan-Arrowood** [comment]
> Sorry for asking this again, but I have really hard time getting this. :) 

No worries thank you for working through this with me! 

> You are suggesting is to NOT add a prettyPrint as a Fastify's factory option in the typings and to let the options being injected by Fastify user?

Yes

> Are we going to remove other typings from here

No

> If not, why some options are "integrated" and others must be provided by passing explicitly Pino typings?

Because of the idea of _minimal viable api_. We should provide a valid, _minimal_ version of the logger types that is easy for us to maintain and will enable a basic TypeScript fastify user to have logging support in their app. If the user needs more than the provided basic api they need to install and inject the Pino types themselves. 

---

My proposal here is already implemented as far as I know. It may need some small changes as well as better documentation. If others disagree with the way it is now then that is okay and we can continue having this discussion

File: types/logger.d.ts
Comment (jsumners) [reply types/logger.d.ts:4]: > I don't like including types that are not under our control (from `@types/pino`) to the fastify dependency tree.
> @mcollina what do you think? What about a PR to pino itself to add typings? @jsumners any thoughts?

As you've learned, I 1. have these TS issues muted and 2. I have opinions. Opinions you will not appreciate.

Comment (fox1t) [reply types/logger.d.ts:4]: @jsumners I appreciate very much your opinions actually. Even more, if your opinions are different from mine. That's why I tagged you here and why I ask your point of view whenever I can. 

However, if you don't want to be tagged on TS issues, I am not going to tag you in the future.

Comment (jsumners) [reply types/logger.d.ts:4]: That would be fine. I cannot provide any useful input for TS related things. If you want the full story, read through https://github.com/fastify/fastify/issues/649

**Comment by mcollina** [comment]
> My explanation here is already implemented as far as I know. It may need some small changes as well as better documentation. If others disagree with the way it is now then that is okay and we can continue having this discussion

I think this is the most important part of this discussion. I don't understand why this PR is needed at all. Can we ease this scenario?

**Comment by fox1t** [comment]
@mcollina this is the first time that an option is more complex than a single value: prettyPrint could be an object. To avoid duplicating its types in fastify, this PR tried to import pino's typings and to use it here. It is the same as we do, for example, in the `fastify-helmet` type definitions. The difference is that `helmet` is shipping its types. 

One possible way of simplifying it is to duplicate in this repo also the prettyPrint object types.

**Comment by mcollina** [comment]
> One possible way of simplifying it is to duplicate in this repo also the prettyPrint object types.

Let's do this.

**Comment by fox1t** [comment]
> > One possible way of simplifying it is to duplicate in this repo also the prettyPrint object types.
> 
> Let's do this.



@agramian You can copy the object from here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/72c9bd83316bd31e93ab86d64ddf598d922f33cd/types/pino/index.d.ts#L514

File: types/logger.d.ts
Comment (fox1t) [reply types/logger.d.ts:4]: I've just finished reading it. That issue was a really long ride.

Comment (jsumners) [reply types/logger.d.ts:4]: Indeed.

**Comment by agramian** [comment]
> > > One possible way of simplifying it is to duplicate in this repo also the prettyPrint object types.
> > 
> > 
> > Let's do this.
> 
> @agramian You can copy the object interface from here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/72c9bd83316bd31e93ab86d64ddf598d922f33cd/types/pino/index.d.ts#L514

Ok I lifted the interface and placed it just above the `FastifyLoggerOptions` interface where it is used.   Let me know if you prefer the new interface to be renamed or relocated as I'm not sure what your preferences are.

[A tweet relevant to this whole discussion.](https://twitter.com/iamdevloper/status/397664295875805184) 😂

**Comment by Ethan-Arrowood** [comment]
Please add a non-jsdoc comment above the interface that says where it was copied from. Additionally, I think now is the best time for us to add a comment at the top of the logger file that summarizes the current decision on why we don't include @types/pino in the dev deps.

**Comment by agramian** [comment]
I added the comment referencing the copy source as well as the summary of the decision as best I could understand it from your guys' discussion.

**Review by Ethan-Arrowood — APPROVED** [review]
I love this. Great work on the comment at the top of the file ❤️

Thank you all for collaborating and discussing this issue!

**Review by mcollina — APPROVED** [review]
lgtm

**Review by L2jLiga — APPROVED** [review]
LGTM!

**Comment by mcollina** [comment]
@fox1t @kibertoad @Eomm are there any outstanding concerns before we land this?

**Review by fox1t — APPROVED** [review]
LGTM

**Comment by fox1t** [comment]
> @fox1t @kibertoad @Eomm are there any outstanding concerns before we land this?

This is perfectly fine. If the pino public interface will ever change we will update these types. 👌

**Review by Eomm — APPROVED** [review]

**Comment by kibertoad** [comment]
refs #2395

**Pull request merged**
